gateway/src/apicast/http_proxy.lua

```diff
@@ -82,15 +98,61 @@ local function absolute_url(uri)
     )
 end
 
-local function forward_https_request(proxy_uri, proxy_auth, uri, skip_https_connect)
-    -- This is needed to call ngx.req.get_body_data() below.
-    ngx.req.read_body()
+local function handle_expect()
+    local expect = ngx.req.get_headers()["Expect"]
+    if type(expect) == "table" then
+        expect = expect[1]
+    end
 
-    local request = {
-        uri = uri,
-        method = ngx.req.get_method(),
-        headers = ngx.req.get_headers(0, true),
-        path = format('%s%s%s', ngx.var.uri, ngx.var.is_args, ngx.var.query_string or ''),
+    if expect and expect:lower() == "100-continue" then
+        ngx.status = 100
+        local ok, err = ngx_send_headers()
+
+        if not ok then
+            return nil, "failed to send response header: " .. (err or "unknown")
+        end
+
+        ok, err = ngx_flush(true)
+        if not ok then
+            return nil, "failed to flush response header: " .. (err or "unknown")
+        end
+    end
+end
+
+local function forward_https_request(proxy_uri, uri, proxy_opts)
+    local body, err
+    local sock
+    local opts = proxy_opts or {}
+    local req_method = ngx_get_method()
+    local encoding = ngx.req.get_headers()["Transfer-Encoding"]
+    local is_chunked = encoding and encoding:lower() == "chunked"
+
+    if http_methods_with_body[req_method] then
+      if opts.request_unbuffered and ngx_http_version() == 1.1 then
+        local _, err = handle_expect()
```

Comment:
I wonder if `Expect` needs to be handled for when buffering is enabled

Reply:
Actually, I do not think we should be doing this. The lua-resty-http lib is doing that for us. WDYT?

Reply:
lib lua-resty-http is a client library and it handles the Expect returned from the server, while we are acting as a server here and need to process the Expect header from the client.

I will run more tests

Reply:
I think that when buffered, APIcast should protect upstream and should handle the `Expect: 100-Continue`. That is, it is the apicast who returns HTTP Response `100 Continue` and then consumes the body before opening the connection to upstream.

When unbuffered, the `Expect: 100-Continue` must be propagated to upstream and it is upstream who decides on responding HTTP Response `100 Continue`. And the lua resty http lib is handling that already, i.e. waiting for the `100 Continue` from upstream before (reading the request body) and sending the body.

I learned, if I am not wrong, that the current `master` branch is actually doing unbuffered approach when there is `Expected: 100 Continue` header. Because it opens connection to upstream and waits for `100 Continue` from upstream, and then it starts reading (from a body reader callback) the request body and sends it. Data is being sent to upstream as soon as it is available. Do you see this behavior as I do?

Reply:
I heave re-written the message above. In case you have read it previosly, please re-read it again :pray:

Reply:
I'm a bit confused here. I haven't read the openresty code but do you mean `ngx.req.read_body()` will send `100 Continue` downstream? Doesn't that also mean that APIcast returns `100 Continue` to the downstream application before establishing the upstream connection?

Regarding the 400, please correct me if I'm wrong, but I think the only case where the upstream server returns this error is if there is data in the request body. In my head the flow will be as follow

```
client -> Expect: 100-Continue -> upstream -> 100 Continue -> client
client -> start sending body -> upstream read body -> return 400
```

Reply:
>  I haven't read the openresty code but do you mean ngx.req.read_body() will send 100 Continue downstream?

Yes!

> Doesn't that also mean that APIcast returns 100 Continue to the downstream application before establishing the upstream connection?

Exactly (when buffered mode is on)

Reply:
From this nginx thread https://mailman.nginx.org/pipermail/nginx/2021-May/060643.html. I think nginx does not handle this well either

How about we send back error response, discard the body and close the socket?

Reply:
> How about we send back error response, discard the body and close the socket?

It's aggressive, but can be a way out.